clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1218,17 +1218,20 @@ public void assign(Collection<TopicPartition> partitions) {
                 throw new IllegalStateException("Consumer is not subscribed to any topics or assigned any partitions");
             }
 
-            // poll for new data until the timeout expires
             do {
                 client.maybeTriggerWakeup();
 
                 if (includeMetadataInTimeout) {
-                    // try to update assignment metadata BUT do not need to block on the timer,
-                    // since even if we are 1) in the middle of a rebalance or 2) have partitions
-                    // with unknown starting positions we may still want to return some data
-                    // as long as there are some partitions fetchable; NOTE we always use a timer with 0ms
-                    // to never block on completing the rebalance procedure if there's any
-                    updateAssignmentMetadataIfNeeded(time.timer(0L));
+                    // try to update assignment metadata BUT do not need to block on the timer if we still have
+                    // some assigned partitions, since even if we are 1) in the middle of a rebalance
+                    // or 2) have partitions with unknown starting positions we may still want to return some data
+                    // as long as there are some partitions fetchable; NOTE we do not block on rebalancing to complete
+                    // if there's one pending if we still have some fetchable partitions
+                    if (subscriptions.fetchablePartitions(tp -> true).isEmpty()) {
```

Comment:
This call to `fetchablePartitions` is unfortunate. It is an entire pass over all the assigned partitions on every poll. Is there any way it can be avoided?

Reply:
Agree, I can think of caching a boolean to avoid it.